# Week 2A: Introduction to PuLP
## Solving a Linear Program (LP) in Python with PuLP

<font color='blue'><b>Goals of this notebook:</b></font>
Learn the basic commands to solve an LP with Python, including `LpProblem` and `LpVariable`.

<font color='blue'><b>Python packages required:</b></font>
PuLP

<font color='blue'><b>Additional resources:</b></font> 
More on PuLP: https://pypi.org/project/PuLP/#description. <br> Additional examples: https://coin-or.github.io/pulp/CaseStudies/a_blending_problem.html.


In its most general form, a linear program looks like

$$
\begin{array}{rcl}
    \max & c^\intercal x\\
    \text{s.t.}& Ax &\le& b\\
                     & Bx &=& d\\
                     & Cx &\ge& f\\
                     & x & \in & \mathbb{R}^n.
\end{array}
$$

There are three components to an LP: the variables $x$, the objective $\max ~ c^\intercal x$, and the constraints $Ax \le b$, $Bx = d$, and $Cx \ge f$.
In order to code these using Python, we can follow the following six steps:

<b>Step 1: Load PuLP.</b> Import Python's toolbox PuLP for solving LPs.

<b>Step 2: Initialize an "empty" linear program.</b>
Intuitively, an empty linear program is Python's version of a sheet of paper on which we write the variables, the objective, and the constraints.

<b>Step 3: Add the variables $x$.</b>
We add the three components of an LP (the variable, the objective function, and the constraints) to the empty linear program created in the last step.

<b>Step 4: Add the objective function </b> $\max ~ c^\intercal x$.

<b>Step 5: Add the constraints </b> $Ax \le b, ~Bx = d$, and $Cx \ge f$.

<b>Step 6: Solve the LP and print the results.</b>

The following example will introduce the basic Python commands needed in <b>Steps 1-6</b>.

## Example: Alice's Farm
Alice wants to build a farm to produce corn. She can spend CHF 1000.
It costs Alice CHF 3 to produce one kilogram of corn, and she can sell it for CHF 7.
Alice can also buy additional farmland at a cost of CHF 100 per acre, and each acre can only grow 30 kilograms of corn.
How many acres and kgs of corn should she buy to maximize profit?

Here is a model of Alice's problem.
    
$$
\begin{array}{rlcl}
\max & 7 \times (\text{corn produced})	\\
\text{s.t.} &  \text{corn produced} &\le& 30 \times (\text{acres purchased})\\
         & 3 \times (\text{corn produced}) + 100 \times (\text{acres purchased}) &\le& 1000\\
         &0 \le \text{corn produced}\\
         &0 \le \text{acres purchased}.
\end{array}
$$

#### Step 1: Loading PuLP

Run the following code block to import PuLP (you do not need to know how this code block works). 

*Note: One way to run the code is to click in the box below and press the 'Run' button. 
Another way is to click in the box and press 'Shift + Enter'.*

In [1]:
# Import the necessary LP toolbox from Python
from pulp import *

#### Step 2: Creating an empty linear program

The function `LpProblem` creates an empty linear program, and it has three parts:
   
* `"Alice's_Farm"` - This is the name displayed when we print the linear program. You can choose any other name you like. 
   
* `LpMaximize` - This makes the linear program a maximization problem.

* `my_LP` - This is the name that the code uses to identify our LP. You can choose any other name you like. 

Run the following code block to create our linear program.

In [2]:
# Create an empty linear program
my_LP = LpProblem("Alice's_Farm", LpMaximize)

#### Step 3: Adding the variables

We need to create two variables: one for the amount of corn produced and one for the number of acres purchased. 
Consider the variable for the amount of corn produced. 
The function to create this variable is `LpVariable` and it has three components:

* `"Corn_produced"` - This is the name displayed when we print the LP. You can choose any other name you like. 

* `"corn"` - This is the name that the code uses to identify this variable. You can choose any other name you like. 

* `lowBound = 0` - This guarantees that the variable is lower bounded by 0 (i.e., that is it nonnegative). 

*Note: The command `lowBound = 0` creates the inequality $\text{corn} \ge 0$. 
Alternatively, we can add this as a constraint in <b>Step 5</b>. 
This type of constraint is very common, so PuLP has commands to add them quickly.*

The `LpVariable` function can also be used to create a variable for the number of acres purchased.

Run the following code block to add our variables.

In [3]:
# Create the variables
corn = LpVariable("Corn_produced", lowBound=0)
acres = LpVariable("Acres_purchased", lowBound=0)

#### Step 4: Adding the objective function

In <b>Step 2</b> that we created `my_LP` to be a maximization problem. 
Therefore, we only need to add the objective function $c^\intercal x$.
The objective function for our problem is $7 \times \text{(corn produced)}$.
Using our variables from <b>Step 3</b>, the objective function becomes `7*corn`.
We use the command `+=` to add this to `my_LP`.

Run the following code block to add this objective to `my_LP`.   
    

In [4]:
# Add the objective function
my_LP += 7*corn

#### Step 5: Adding the constraints

The first constraint is $\text{corn produced} \le 30 \times (\text{acres purchased})$.
Using our variables this becomes `corn <= 30* acres`.
We use the command `+=` to add this to `my_LP`.

Run the following code block to create this constraint and add it to the `my_LP`. 

In [5]:
# Add the first constraint
my_LP += corn <= 30*acres

The second constraint is $3 \times (\text{corn produced}) + 100 \times (\text{acres purchased}) \le 1000$.

Add this to our LP by running the following code block.

In [6]:
# Add the second constraint
my_LP += 3*corn + 100*acres <= 1000

#### Step 6: Solving the LP and print the results

Now our linear program is completely built!
The following code block will display the linear program so that you can check it.

In [7]:
# Display our linear program
print(my_LP)

Alice's_Farm:
MAXIMIZE
7*Corn_produced + 0
SUBJECT TO
_C1: - 30 Acres_purchased + Corn_produced <= 0

_C2: 100 Acres_purchased + 3 Corn_produced <= 1000

VARIABLES
Acres_purchased Continuous
Corn_produced Continuous



*Note: After running the `print(my_LP)` command, the variables displayed will not explicitly state that the variables are lower bounded by zero. Also, the variables will be displayed as `Continuous`. This can be ignored for now, but we will return to this when we consider discrete decision problems.*

The following code block solves our linear program.

In [9]:
# Solve the linear program
my_LP.solve()

Welcome to the CBC MILP Solver 
Version: 2.10.8 
Build Date: May  6 2022 

command line - cbc /var/folders/wb/nsj1jwbn6pg7v8z5l7scpkw80000gn/T/f1b4edd691584505ba5e22aa2390464a-pulp.mps max timeMode elapsed branch printingOptions all solution /var/folders/wb/nsj1jwbn6pg7v8z5l7scpkw80000gn/T/f1b4edd691584505ba5e22aa2390464a-pulp.sol (default strategy 1)
At line 2 NAME          MODEL
At line 3 ROWS
At line 7 COLUMNS
At line 13 RHS
At line 16 BOUNDS
At line 17 ENDATA
Problem MODEL has 2 rows, 2 columns and 4 elements
Coin0008I MODEL read with 0 errors
Option for timeMode changed from cpu to elapsed
Presolve 0 (-2) rows, 0 (-2) columns and 0 (-4) elements
Empty problem - 0 rows, 0 columns and 0 elements
Optimal - objective value 1105.2632
After Postsolve, objective 1105.2632, infeasibilities - dual 0 (0), primal 0 (0)
Optimal objective 1105.263158 - 0 iterations time 0.002, Presolve 0.00
Option for printingOptions changed from normal to all
Total time (CPU seconds):       0.00   (Wallclock 

1

If everything went well, then the output should be `1`. 
It does not mean that the solution to `my_LP` equals 1. Instead, the output indicates that the program was executed successfully without any errors. 
However, we (Alice and us) would like to know the optimal objective value and the optimal values of `corn` and `acres`.

*Note: The possible output values are `-3,-2,-1,0` or `1`, indicating e.g. if the problem was solved to optimality (`0`), found to be infeasible (`-1`) or unbounded (`-2`). You my check the documentation for details: https://coin-or.github.io/pulp/technical/constants.html?highlight=constants#pulp.constants.LpStatus.*


The optimal value of `corn` can be accessed using `corn.value()`.
Similarly, the optimal value of `acres` can be accessed using `acres.value()`.
The optimal objective value of `my_LP` can be accessed using `value(my_LP.objective)`.

*Note: The character `%.2f` in the code below is a formatting tool for rounding a decimal to two places. It is not important for this tutorial to know how this formatting works. For more details, see https://docs.python.org/2/library/string.html.*

Run the following code block to display the optimal values from `my_LP`. 

In [10]:
# Print the optimal value and the variables values

opt_corn = corn.value()
print(f'Alice should produce {opt_corn:.2f} kilograms of corn.')

opt_acres = acres.value()
print(f'Alice should purchase {opt_acres:.2f} acres of land.')

opt_val = value(my_LP.objective)
print(f'Alice will have a profit of CHF {opt_val:.2f}.')

Alice should produce 157.89 kilograms of corn.
Alice should purchase 5.26 acres of land.
Alice will have a profit of CHF 1105.26.


Congratulations! We have successfully solved our first LP. If everything was run correctly, then the output should be

    Alice should produce 157.89 kilograms of corn.
    Alice should purchase 5.26 acres of land.
    Alice will have a profit of CHF 1105.26.

#### Conclusions

There are six main steps to solving a linear program. 
Moreover, the basic commands that we learned are already enough to solve many optimization problems. See https://coin-or.github.io/pulp/CaseStudies/a_blending_problem.html for another example.